### DISCUSSION TITLE
RS41 running with solar cells only,  needs additional support

### DISCUSSION THREAD

**Discussion by whallmann** [mikaelnousiainen/RS41ng#63] [Ideas]
Hi,
i am succeeded in running a rs41 sonde only with powered solar cell arrays.  This is the start of some investigations to send the rs41 on longer flys like this lightweight items U4B etc. But the rs41 has some peculiarities.
If there is an external C (1000 uF) buffering the solar power of 4,2 Volts the rs41 starts after a delayed VCC connection after the C is filled up. So long, so good.
The main problem to discuss is this: if the solar cell array gets first sunnshine on sunset, the solar array deliveres to low energy. The RS41 hangs in a start loop (led is permanently on). This look like the main programm will not start. 
So in my oppinion there should be a chance to make a total reset of the cpu if gps is not ready after we say 5 Minutes.
This should be possible by CPU interrupt?!  
What i see here is following: i programmed a PIP after every 30 seconds. This gets the TX led high for some milliseconds. So i think the TX queue is somehow running, while the main LED shows still permanently lit.
If there is some code running in the rs41, there should be a way to reset something - possible the hole cpu.  
So after some Resets there is enough power to get the rs41 starting up normally.
This is a challenge, isn´t it?   

PS: i powered the rs41 not over the powermanagement chip, i soldered the VCC direct to the smd item to start CPU without pushing this button on the pcb.

Internet says: NVIC_SystemReset() is a possible chance to make a cold start

**Comment by whallmann** [comment]
Maybe this **SystemReset** will help at other moments. Some time the gps will not come up, because i have to go outside to have gps receiption. I feel that gps will never come up after some short time intervall, even gps is available after the this time.  
On this status, i need to restart the sonde again, then it comes up after less then 30 seconds. It looks like, there is a timeout the programm will never get a gps response from the chip, so also a SystemReset will help if we waiting long enough.

**Reply by mikaelnousiainen** [reply]
You could indeed experiment with tracking the time spent when there is no GPS fix and call reset after a certain threshold is reached.

However, if the red LED stays constantly lit, it means that the MCU is not executing code (locked up because of undervoltage, possibly?) or there is some condition in early setup of MCU clock or peripherals which is never met -- meaning MCU is not behaving correctly. The code will boot and start transmitting always, regardless of GPS lock.

You might need some additional circuitry for handling undervoltage... and IIRC there is some kind of watchdog functionality available in the STM32F100-series MCUs that could reset the processor. It might be worth looking into that!

Also, you mentioned you are not using the RS41 power management circuitry -- how do you downcovert 4.2 V to 3.3V? 4.2 V is way too high for the MCU and other chips involved.

**Reply by whallmann** [reply]
Hi Mikael,  last thing first:  the 4,2 Volt are powered direct - there is no downconverting. Sonde comes up 100% correct ;-) 
Interessting is the situation with the MCU. You agree there is some code running, sounds helpfull for more experiments.

**Reply by mikaelnousiainen** [reply]
4.2 V is too much even if it seems to work. It may damage the equipment.

Also, I was indicating that code may or may NOT be running -- there are some loops that require MCU state changes in the boot procedure, which -- if the MCU is not behaving correctly -- lead to endless loops even if code is being executed. In any case, a full reset should be done to attempt to fix it. If the MCU behaves as it should, the red LED turns off and transmissions start. So something is going seriously wrong during boot.

**Reply by iz2fly** [reply]
The TPS61200 Boost Converter is mounted at the input on the battery side, the voltage for the RS41 remains constant at 3.75V, even when powering the probe at 5V the probe always works at 3.75V, the same with a lower power supply of up to 2V. The lowest consumption occurs at 3.3V.
![RS41_Watt_01](https://github.com/mikaelnousiainen/RS41ng/assets/56764775/716448ea-5bbd-4d74-b1f0-d742af842f17)

**Reply by iz2fly** [reply]
Perhaps we would need a circuit that turns off the sonde at 2V and turns it back on at 3.3V.
I did it this way with my beacon powered by solar panels and 12V battery.
This way switching on and off would be error-free.
I don't know if this can be done by the sonde itself.

**Reply by whallmann** [reply]
Meanwhile i got some  results. First i added a counter if there comes no GPSfix after a timeout . So it will reboot by system reset. This works good i would say after manny hours of tests.
Second i mounted solar panels 3,8 Volt to one of the smd items bypassed the regulator. If the voltage is low the gpsinit fails and got into Infiniti loop. I breaked the loop by resetting with an system reset after 10 retrys. Since this mod the software comes up regulary if the voltage climbs over a level of 0,9 Volt. I am happy not have to add more items like startup regulators or reed relays some guys using.
More Infos in my GitHub fork on whallmann/RS41ng. See README there.
